platform/cog-platform-gtk4.c

```diff
@@ -0,0 +1,651 @@
+    GtkShortcutTrigger* activate_entry_trigger = gtk_shortcut_trigger_parse_string("<Control>l");
+    GtkShortcut* activate_entry_shortcut = gtk_shortcut_new(activate_entry_trigger, activate_entry_action);
+    gtk_shortcut_controller_add_shortcut(shortcut_controller,
+        activate_entry_shortcut);
+
+    gtk_widget_add_controller(window->gtk_window,
+        GTK_EVENT_CONTROLLER(shortcut_controller));
+
+    GtkWidget* box = gtk_box_new(GTK_ORIENTATION_VERTICAL, FALSE);
+    gtk_window_set_child(GTK_WINDOW(window->gtk_window), box);
+    gtk_box_append(GTK_BOX(box), window->gl_drawing_area);
+
+    gtk_widget_show(window->gtk_window);
+}
+
+static void on_export_egl_image(void* userdata,
+    struct wpe_fdo_egl_exported_image* image)
+{
+    struct platform_window* window = userdata;
+
+    window->current_image = image;
+    gtk_gl_area_queue_render(GTK_GL_AREA(window->gl_drawing_area));
+}
+
+static void setup_fdo_exportable(struct platform_window* window)
+{
+    g_assert_nonnull(window);
+
+    static const struct wpe_view_backend_exportable_fdo_egl_client client = {
+        .export_fdo_egl_image = on_export_egl_image,
+    };
+
+    window->exportable = wpe_view_backend_exportable_fdo_egl_create(
+        &client, window, (uint32_t)DEFAULT_WIDTH, (uint32_t)DEFAULT_HEIGHT);
+    g_assert_nonnull(window->exportable);
+
+    window->view_backend = webkit_web_view_backend_new(
+        wpe_view_backend_exportable_fdo_get_view_backend(window->exportable),
+        (GDestroyNotify)wpe_view_backend_exportable_fdo_destroy,
+        window->exportable);
+    g_assert_nonnull(window->view_backend);
+}
+
+gboolean cog_platform_plugin_setup(CogPlatform* platform,
+    CogShell* shell G_GNUC_UNUSED,
+    const char* params, GError** error)
+{
+    g_assert_nonnull(platform);
+
+    wpe_loader_init("libWPEBackend-fdo-1.0.so");
+    if (!gtk_init_check()) {
+        g_set_error_literal(error, COG_PLATFORM_EGL_ERROR, 0,
+            "GTK initialization failed");
+        return FALSE;
+    }
+
+    setup_window(&win);
+    setup_fdo_exportable(&win);
+    return TRUE;
+}
+
+void cog_platform_plugin_teardown(CogPlatform* platform)
+{
+    g_assert_nonnull(platform);
+}
+
+WebKitWebViewBackend* cog_platform_plugin_get_view_backend(
+    CogPlatform* platform, WebKitWebView* related_view, GError** error)
+{
+    g_assert_nonnull(platform);
+    g_assert_nonnull(win.view_backend);
+    return win.view_backend;
+}
+
+static void on_title_change(WebKitWebView* view, GParamSpec* pspec,
+    gpointer user_data)
+{
+    g_autofree gchar* title;
+    g_object_get(view, "title", &title, NULL);
+    g_autofree gchar* win_title = g_strdup_printf("Cog - %s", title);
+    gtk_window_set_title(GTK_WINDOW(win.gtk_window), win_title);
+}
+
+static void on_uri_change(WebKitWebView* view, GParamSpec* pspec,
+    gpointer user_data)
+{
+    g_autofree gchar* uri;
+    g_object_get(view, "uri", &uri, NULL);
```

Comment:
This could use `webkit_web_view_get_uri()`, for simplicity.

```suggestion
    const char* uri = webkit_web_view_get_uri (view);
```